submissions/VBystrov/task_js_oop/app.js

```diff
@@ -0,0 +1,169 @@
+  return Math.floor(Math.random() * (max - min) + min);
+}
+
+const Enemy = function () {
+  this.sprite = 'images/enemy-bug.png';
+  Resources.load(this.sprite);
+};
+
+Enemy.prototype.update = function (dt) {
+  const resetLine = 640;
+  this.x += this.speed * dt;
+  if (this.x > resetLine) {
+    this.startPosition();
+  }
+};
+
+Enemy.prototype.startPosition = function () {
+  this.x = randomMinMax(enemyMinStartX, enemyMaxStartX);
+  this.y = randomMinMax(enemyMinStartY, enemyMaxStartY);
+  this.speed = randomMinMax(enemyMinSpeed, enemyMaxSpeed);
+};
+
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const Player = function () {
+  this.speed = playerSpeed;
+  this.sprite = 'images/char-horn-girl.png';
+  Resources.load(this.sprite);
+};
+
+Player.prototype.update = function (dx = 0, dy = 0) {
+  const minX = -10;
+  const maxX = 420;
+  const minY = -10;
+  const maxY = 400;
+
+  this.x += dx * this.speed;
+  this.y += dy * this.speed;
+  if (this.x < minX) {
+    this.x = minX;
+  }
+  if (this.x > maxX) {
+    this.x = maxX;
+  }
+  if (this.y < minY) {
+    this.y = minY;
+  }
+  if (this.y > maxY) {
+    this.y = maxY;
+  }
+};
+
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (direction) {
+  const d = 1;
+  let dx = 0;
+  let dy = 0;
+  switch (direction) {
+    case 'left':
+      dx = -d;
+      break;
+    case 'up':
+      dy = -d;
+      break;
+    case 'right':
+      dx = d;
+      break;
+    case 'down':
+      dy = d;
+      break;
+    default:
+      break;
+  }
+  this.update(dx, dy);
+};
+
+Player.prototype.checkCollisions = function (enemies) {
+  // coordinates collision
+  const minX = 18;
+  const maxX = 84;
+  const minY = 85;
+  const maxY = 137;
+  return enemies.some(
+    function (enemy) {
+      const dx = this.x - enemy.x;
+      const dy = this.y - enemy.y;
+      const checkPointX = dx > 0 ? this.x + minX : this.x + maxX;
+      const checkPointY = dy > 0 ? this.y + minY : this.y + maxY;
+      if (
+        checkPointX >= enemy.x + minX &&
+        checkPointX <= enemy.x + maxX &&
+        checkPointY >= enemy.y + minY &&
+        checkPointY <= enemy.y + maxY
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    }.bind(this)
+  );
+};
+
+Player.prototype.startPosition = function () {
+  this.x = playerStartX;
+  this.y = playerStartY;
+};
+
+Player.prototype.inWater = function () {
+  const finishLine = -10;
+  return this.y <= finishLine;
+};
+
+function checkCollisions() {
+  if (player.checkCollisions(allEnemies) || player.inWater()) {
+    allEnemies.forEach(function (enemy) {
+      enemy.startPosition();
+    });
+    player.startPosition();
+  }
+}
+
+const allEnemies = [];
+for (let i = 0; i < numEnemies; i++) {
+  allEnemies.push(new Enemy());
+  allEnemies[i].startPosition();
```

Comment:
Consider calling `.startPosition` from inside constructor since this is smth called anyway every time when an enemy gets created.
`startPosition` reads ambiguously. If `start` here is a verb then it reads "start to position (itself?)". If it is a noun then we're lacking a verb in method name.

Reply:
Done.